File: Java/01. Introduction/Practice02.java
/**
 * Practice 02: Find Min & Max Salary
 * Task: Read employee salary data from a CSV file and display the employee
 *       with the highest and lowest salary.
 *
 * How to compile and run:
 *   javac Practice02.java
 *   java Practice02
 *
 * Expected output:
 *   Max: Hasan   56000
 *   Min: Rafi    34000
 *
 * NOTE: The data file is at: ../../data/salarysheet.csv
 *       Make sure you run this program from the Java/01. Introduction/ directory.
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Practice02 {

    /**
     * Inner class: represents one employee with name and salary.
     */
    static class Employee {
        String name;
        double salary;

        Employee(String name, double salary) {
            this.name = name;
            this.salary = salary;
        }
    }

    public static void main(String[] args) {
        String fileName = "../../data/salarysheet.csv";

        // Step 1: Read data from file
        List<Employee> salaryList = readDataFromCSVFile(fileName);

        if (salaryList == null || salaryList.isEmpty()) {
            System.out.println("No employee data found in the file.");
            return;
        }

        // Step 2: Find max and min salary employees
        Employee maxEmployee = findMaxSalaryWithEmployeeName(salaryList);
        Employee minEmployee = findMinSalaryWithEmployeeName(salaryList);

        // Step 3: Display the results
        System.out.printf("Max: %-10s %.0f%n", maxEmployee.name, maxEmployee.salary);
        System.out.printf("Min: %-10s %.0f%n", minEmployee.name, minEmployee.salary);
    }

    /**
     * Reads employee names and salaries from a tab-separated file.
     * Each line should have: EmployeeName <tab> Salary
     */
    public static List<Employee> readDataFromCSVFile(String fileName) {
        List<Employee> employees = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) continue; // skip blank lines

                // Split by tab first, then fallback to comma
                String[] parts = line.split("\t");
                if (parts.length < 2) {
                    parts = line.split(",");
                }
                if (parts.length < 2) continue; // skip malformed lines

                String name = parts[0].trim();
                double salary = Double.parseDouble(parts[1].trim());
                employees.add(new Employee(name, salary));
            }
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }

        return employees;
    }

    /**
     * Returns the employee with the highest salary.
     */
    public static Employee findMaxSalaryWithEmployeeName(List<Employee> employees) {
        Employee max = employees.get(0);
        for (Employee emp : employees) {
            if (emp.salary > max.salary) {
                max = emp;
            }
        }
        return max;
    }

    /**
     * Returns the employee with the lowest salary.
     */
    public static Employee findMinSalaryWithEmployeeName(List<Employee> employees) {
        Employee min = employees.get(0);
        for (Employee emp : employees) {
            if (emp.salary < min.salary) {
                min = emp;
            }
        }
        return min;
    }
}

File: C#/01. Introduction/Practice02.cs
/**
 * Practice 02: Find Min & Max Salary
 * Task: Read employee salary data from a CSV file and display the employee
 *       with the highest and lowest salary.
 *
 * How to run:
 *   Create a Console App, paste this code into Program.cs, and run.
 *   OR use: dotnet-script Practice02.cs
 *
 * Expected output:
 *   Max: Hasan   56000
 *   Min: Rafi    34000
 *
 * NOTE: The data file is at: ../../data/salarysheet.csv
 *       Make sure the working directory is C#/01. Introduction/ when running.
 */

using System;
using System.Collections.Generic;
using System.IO;

class Practice02
{
    /**
     * Inner class: represents one employee with name and salary.
     */
    class Employee
    {
        public string Name { get; set; }
        public double Salary { get; set; }

        public Employee(string name, double salary)
        {
            Name = name;
            Salary = salary;
        }
    }

    static void Main(string[] args)
    {
        string fileName = "../../data/salarysheet.csv";

        // Step 1: Read data from file
        List<Employee> salaryList = ReadDataFromCSVFile(fileName);

        if (salaryList == null || salaryList.Count == 0)
        {
            Console.WriteLine("No employee data found in the file.");
            return;
        }

        // Step 2: Find max and min salary employees
        Employee maxEmployee = FindMaxSalaryWithEmployeeName(salaryList);
        Employee minEmployee = FindMinSalaryWithEmployeeName(salaryList);

        // Step 3: Display the results
        Console.WriteLine($"Max: {maxEmployee.Name,-10} {maxEmployee.Salary:F0}");
        Console.WriteLine($"Min: {minEmployee.Name,-10} {minEmployee.Salary:F0}");
    }

    /**
     * Reads employee names and salaries from a tab-separated file.
     * Each line should have: EmployeeName TAB Salary
     */
    static List<Employee> ReadDataFromCSVFile(string fileName)
    {
        List<Employee> employees = new List<Employee>();

        try
        {
            string[] lines = File.ReadAllLines(fileName);
            foreach (string line in lines)
            {
                string trimmed = line.Trim();
                if (string.IsNullOrEmpty(trimmed)) continue; // skip blank lines

                // Split by tab first, then fallback to comma
                string[] parts = trimmed.Split('\t');
                if (parts.Length < 2)
                    parts = trimmed.Split(',');
                if (parts.Length < 2) continue; // skip malformed lines

                string name = parts[0].Trim();
                double salary = double.Parse(parts[1].Trim());
                employees.Add(new Employee(name, salary));
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error reading file: {ex.Message}");
        }

        return employees;
    }

    /**
     * Returns the employee with the highest salary.
     */
    static Employee FindMaxSalaryWithEmployeeName(List<Employee> employees)
    {
        Employee max = employees[0];
        foreach (Employee emp in employees)
        {
            if (emp.Salary > max.Salary)
                max = emp;
        }
        return max;
    }

    /**
     * Returns the employee with the lowest salary.
     */
    static Employee FindMinSalaryWithEmployeeName(List<Employee> employees)
    {
        Employee min = employees[0];
        foreach (Employee emp in employees)
        {
            if (emp.Salary < min.Salary)
                min = emp;
        }
        return min;
    }
}
